firstpartydata/first_party_data.go

```diff
@@ -0,0 +1,376 @@
+	newSite.Page = fpdConfigSite.Page
+	newSite.Search = fpdConfigSite.Search
+	newSite.Keywords = fpdConfigSite.Keywords
+
+	if len(fpdConfigSite.Ext) > 0 {
+		if len(originalSite.Ext) > 0 {
+			newSite.Ext, err = jsonpatch.MergePatch(originalSite.Ext, fpdConfigSite.Ext)
+		} else {
+			newSite.Ext = fpdConfigSite.Ext
+		}
+	}
+
+	return newSite, err
+}
+
+func resolveApp(fpdConfigApp *openrtb2.App, bidRequestApp *openrtb2.App, globalFPD map[string][]byte, openRtbGlobalFPD map[string][]openrtb2.Data, hasGlobalFPD bool, bidderName string) (*openrtb2.App, error) {
+
+	if bidRequestApp == nil && fpdConfigApp == nil {
+		return nil, nil
+	}
+
+	if bidRequestApp == nil && fpdConfigApp != nil {
+		return nil, fmt.Errorf("incorrect First Party Data for bidder %s: App object is not defined in request, but defined in FPD config", bidderName)
+	}
+
+	newApp := *bidRequestApp
+	var err error
+
+	if hasGlobalFPD {
+		//apply global fpd if exists
+		if len(globalFPD[appKey]) > 0 {
+			extData := buildExtData(globalFPD[appKey])
+			if len(newApp.Ext) > 0 {
+				newApp.Ext, err = jsonpatch.MergePatch(newApp.Ext, extData)
+			} else {
+				newApp.Ext = extData
+			}
+		}
+		if len(openRtbGlobalFPD[appContentDataKey]) > 0 {
+			if newApp.Content != nil {
+				contentCopy := *newApp.Content
+				contentCopy.Data = openRtbGlobalFPD[appContentDataKey]
+				newApp.Content = &contentCopy
+			} else {
+				newAppContent := &openrtb2.Content{Data: openRtbGlobalFPD[appContentDataKey]}
+				newApp.Content = newAppContent
+			}
+		}
+	}
+
+	if fpdConfigApp != nil {
+		//apply bidder specific fpd if present
+		newApp, err = mergeApps(&newApp, fpdConfigApp)
+	}
+
+	return &newApp, err
+}
+
+func mergeApps(originalApp *openrtb2.App, fpdConfigApp *openrtb2.App) (openrtb2.App, error) {
+
+	var err error
+	newApp := *originalApp
+
+	newApp.Name = fpdConfigApp.Name
+	newApp.Bundle = fpdConfigApp.Bundle
+	newApp.Domain = fpdConfigApp.Domain
+	newApp.StoreURL = fpdConfigApp.StoreURL
+	newApp.Cat = fpdConfigApp.Cat
+	newApp.SectionCat = fpdConfigApp.SectionCat
+	newApp.PageCat = fpdConfigApp.PageCat
+	newApp.Ver = fpdConfigApp.Ver
+	newApp.Keywords = fpdConfigApp.Keywords
+
+	if len(fpdConfigApp.Ext) > 0 {
+		if len(originalApp.Ext) > 0 {
+			newApp.Ext, err = jsonpatch.MergePatch(originalApp.Ext, fpdConfigApp.Ext)
+		} else {
+			newApp.Ext = fpdConfigApp.Ext
+		}
+	}
+
+	return newApp, err
+}
+
+func buildExtData(data []byte) []byte {
+	res := make([]byte, 0, len(data))
+	res = append(res, []byte(`{"data":`)...)
+	res = append(res, data...)
+	res = append(res, []byte(`}`)...)
+	return res
+}
+
+func ExtractBidderConfigFPD(reqExtPrebid openrtb_ext.ExtRequestPrebid) (map[openrtb_ext.BidderName]*openrtb_ext.ORTB2, openrtb_ext.ExtRequestPrebid) {
+
+	fpdData := make(map[openrtb_ext.BidderName]*openrtb_ext.ORTB2)
+
+	//every bidder in ext.prebid.data.bidders should receive fpd data if defined
+	bidderTable := make(map[string]struct{}) //just need to check existence of the element in map
```

Comment:
Consider renaming to something like `bidderUniquenessTracker` so the comment isn't necessary.

Reply:
This is removed in previous refactoring